test/e2e/cluster_deletion.go

```diff
@@ -0,0 +1,357 @@
+			return relevantKinds.Has(u.GetKind())
+		})
+		Expect(err).ToNot(HaveOccurred())
+
+		for i, phase := range input.ClusterDeletionPhases {
+			// Iterate over the graph and find all objects relevant for the phase.
+			// * Objects matching the kind except for Machines (see below)
+			// * Machines which are owned by one of the kinds which are part of the phase
+			allKinds := phase.KindsWithFinalizer.Union(phase.Kinds)
+			allKindsWithoutMachine := allKinds.Clone().Delete("Machine")
+
+			for _, node := range graph {
+				// Only add objects which match a defined kind for the phase.
+				if !allKinds.Has(node.Object.Kind) {
+					continue
+				}
+
+				// Special handling for machines: only add machines which are owned by one of the given kinds.
+				if node.Object.Kind == "Machine" {
+					isOwnedByKind := false
+					for _, owner := range node.Owners {
+						if allKindsWithoutMachine.Has(owner.Kind) {
+							isOwnedByKind = true
+							break
+						}
+					}
+					// Ignore Machines which are not owned by one of the given kinds.
+					if !isOwnedByKind {
+						continue
+					}
+				}
+
+				obj := new(unstructured.Unstructured)
+				obj.SetAPIVersion(node.Object.APIVersion)
+				obj.SetKind(node.Object.Kind)
+				obj.SetName(node.Object.Name)
+				obj.SetNamespace(namespace.GetName())
+
+				phase.objects = append(phase.objects, obj)
+
+				// If the object's kind is part of kindsWithFinalizer, then additionally the object to the finalizedObjs slice.
+				if phase.KindsWithFinalizer.Has(obj.GetKind()) {
+					objectsWithFinalizer = append(objectsWithFinalizer, obj)
+				}
+			}
+
+			// Update the phase in input.ClusterDeletionPhases
+			input.ClusterDeletionPhases[i] = phase
+		}
+
+		// Update all objects in objectsWithFinalizers and add the finalizer for this test to control when these objects vanish.
+		addFinalizer(ctx, input.BootstrapClusterProxy.GetClient(), finalizer, objectsWithFinalizer...)
+
+		// Trigger the deletion of the Cluster.
+		framework.DeleteCluster(ctx, framework.DeleteClusterInput{
+			Cluster: clusterResources.Cluster,
+			Deleter: input.BootstrapClusterProxy.GetClient(),
+		})
+
+		var objectsDeleted, objectsInDeletion []client.Object
+
+		for i, phase := range input.ClusterDeletionPhases {
+			// Objects which previously were in deletion because they waited for our finalizer being removed are now checked to actually had been removed.
+			objectsDeleted = objectsInDeletion
+			// All objects for this phase are expected to get into deletion (still exist and have deletionTimestamp set).
+			objectsInDeletion = phase.objects
+			// All objects which are part of future phases are expected to not yet be in deletion.
+			objectsNotInDeletion := []client.Object{}
+			for j := i + 1; j < len(input.ClusterDeletionPhases); j++ {
+				objectsNotInDeletion = append(objectsNotInDeletion, input.ClusterDeletionPhases[j].objects...)
+			}
+
+			Eventually(func(g Gomega) {
```

Comment:
Wondering if we should rewrite this without gomega asserting panicking at the first issue, but instead collect issues in a list of errors/messages, and print it out if things are not in the state we expect
WDYT?